chips/sam4l/src/adc.rs

```diff
@@ -190,20 +201,136 @@ impl adc::AdcSingle for Adc {
 }
 
 /// Not implemented yet. -pal 12/22/16
-impl adc::AdcContinuous for Adc {
+impl adc::AdcContinuous for Adc { 
+    
+    // Unit is Hz.
     type Frequency = adc::Freq1KHz;
 
-    fn compute_interval(&self, interval: u32) -> u32 {
-        interval
+    // Unit is microsecond for parameter `interval` and return value.
+    fn compute_interval(&self, _interval: u32) -> u32 {
+        compute_interval_freq(_interval, Self::Frequency::frequency())
     }
 
     fn sample_continuous(&self, _channel: u8, _interval: u32) -> ReturnCode {
-        ReturnCode::FAIL
+        let regs: &mut AdcRegisters = unsafe { mem::transmute(self.registers) };
+        self.enabled.set(true);
+        // This logic is from 38.6.1 "Initializing the ADCIFE" of
+        // the SAM4L data sheet
+        // 1. Start the clocks, ADC uses GCLK10, choose to
+        // source it from RC32K (32Khz)
+        unsafe {
+            pm::enable_clock(Clock::PBA(PBAClock::ADCIFE));
+            nvic::enable(nvic::NvicIdx::ADCIFE);
+            if self.current_frequency.get() <= 1000 {
+                scif::generic_clock_enable(scif::GenericClock::GCLK10, scif::ClockSource::RC32K);
+            } else if self.current_frequency.get() <= 10000 {
+                scif::generic_clock_enable(scif::GenericClock::GCLK10, scif::ClockSource::RCSYS);
+            } else {
+                scif::generic_clock_enable(scif::GenericClock::GCLK10, scif::ClockSource::RCFAST);
+            }
```

Comment:
Changing clocks like this is a great idea, with two caveats:

1) I've personally found initialization of the ADC to be quite tricky. I only managed to get it to work for continuous sampling at all by changing the order in which I do things (https://github.com/helena-project/tock/blob/adc-continuous/chips/sam4l/src/adc.rs#L276). The current code you have follows the wording in the datasheet, which you'd really think would work, but apparently not. Even with that changed, I wouldn't be surprised if the ADC starts to have problems when its clocks are changed. Calling a reset and then disabling the module before changing clocks could possibly help. The greater point here though is that while changing clocks based on demand is a good idea, I'm concerned about the amount of engineering work that would be required to make it actually functional versus the potential gains.

2) With analog sampling, we're concerned both about the frequency at which samples are taken and the jitter on an individual sample. Clocks based on an RC oscillator are often far more variable than clocks based on a crystal. Although I don't have any quantitative numbers here, I would be concerned that using the RC clocks to run the ADC would result in high variability from sample to sample.